# Wirtualne terrarium

### Joanna Tustanowska, Marcin Markiewicz
### 27.11.2018, 10.12.2019, Python4Beginners

**Plan na dziś**

**1. Jak zarządzać swoją instalacją Pythona i nie zwariować** `<-------`
    * dlaczego powinniśmy to robić
    * jakie mamy narzędzia i jak działają
    * dobre praktyki
    
**2. Biblioteka requests - HTTP i API w Pythonie**
    * protokół HTTP (co to jest protokół i co to jest HTTP)
    * JSON API
    * jak wypełnić formularz w pythonie?
    * co robić, gdy potrzebne jest zalogowanie się?
    * do poczytania (scrapy, httpie/postman/curl, więcej o API design)
    


# Część 1: 
# Jak zarządzać swoją instalacją Pythona i nie zwariować?


## Wstęp - gdzie mieszka Python?

Przykład - domyślna instalacja Pythona na Linuksie. To tylko niektóre pliki.

```console
$ which python
/bin/python
```

```console
$ ls /usr/lib/python3.7
-rw-r--r--  1 root root 5,5K 10-22 12:41 abc.py
-rwxr-xr-x  1 root root  20K 10-22 12:41 base64.py
...
drwxr-xr-x 88 root root 4,0K 11-16 15:26 site-packages
...
```

* `site-packages` to katalog, do którego trafiają zewnętrzne pakiety Pythona

### Ratunku! używam Windowsa

W konsoli windows na próżno szukać nam komendy `which`, jedną z szybszych i skuteczniejszych możliwości sprawdzenia gdzie jest zainstalowany nasz python, jest jego uruchomienie i zapytanie się gdzie on jest:

In [3]:
import sys
sys.executable


'/home/mmark/.local/share/virtualenvs/p4b-0H8Okv2r/bin/python'

Windows & Mac

## `pip` - narzędzie do instalacji pakietów

`pip` - narzędzie dystrybuowane razem z Pythonem (od wersji 3.4)

`$ pip install [nazwa pakietu]`  / można też: `$ python -m pip install [nazwa pakietu]`

np. `pip install maya` 

Zainstaluje pakiet oraz jego zależności.

Instalacja do `site-packages` użytkownika (nie wymaga uprawnień administratora):

`$ pip install --user [nazwa paczki]`



### Odinstalowanie pakietu

`$ pip uninstall [nazwa pakietu]`

Odinstaluje pakiet, pozostawi jego zależności.


### Lista zainstalowanych pakietów

```console
$ pip freeze
dateparser==0.7.0
humanize==0.5.1
pendulum==1.5.1
python-dateutil==2.7.5
[...]
```

`pip freeze > requirements.txt`

### Instalacja z requirements.txt

`$ pip install -r requirements.txt`


In [2]:
pip freeze

alabaster==0.7.12
attrs==19.3.0
Babel==2.7.0
backcall==0.1.0
bleach==3.1.0
certifi==2019.11.28
chardet==3.0.4
decorator==4.4.1
defusedxml==0.6.0
docutils==0.15.2
entrypoints==0.3
idna==2.8
imagesize==1.1.0
importlib-metadata==1.2.0
ipykernel==5.1.3
ipyparallel==6.2.4
ipython==7.10.1
ipython-genutils==0.2.0
ipywidgets==7.5.1
itsdangerous==1.1.0
jedi==0.15.1
Jinja2==2.10.3
json5==0.8.5
jsonschema==3.2.0
jupyter==1.0.0
jupyter-client==5.3.4
jupyter-console==6.0.0
jupyter-contrib-core==0.3.3
jupyter-contrib-nbextensions==0.5.1
jupyter-core==4.6.1
jupyter-highlight-selected-word==0.2.0
jupyter-latex-envs==1.4.6
jupyter-nbextensions-configurator==0.4.1
jupyterlab==1.2.3
jupyterlab-server==1.0.6
lxml==4.4.2
MarkupSafe==1.1.1
mistune==0.8.4
more-itertools==8.0.2
nbconvert==5.6.1
nbformat==4.4.0
nose==1.3.7
notebook==6.0.2
numpy==1.17.4
packaging==19.2
pandocfilters==1.4.2
parso==0.5.1
pexpect==4.7.0
pickleshare==0.7.5
prometheus-client==0.7.1
prompt-toolkit==3.0.2
ptyprocess==0.6.0
Pygments==2

### Wyszukiwanie pakietów w (domyślnie w PyPI https://pypi.org/)

`$ pip search [nazwa pakietu]`

### Inne cechy

* `pip` zapisuje pobrane pakiety w cache'u, dla szybszej ponownej instalacji
* `pip check` - sprawdzi, czy w środowisku zainstalowane są wszystkie wymagane zależności, oraz czy nie ma konfliktów (np. pakiet A wymaga pakietu C w wersji <1.0, zaś pakiet B wymaga pakietu C w wersji >1.0)
* Dokumentacja `pip` do poczytania: https://pip.pypa.io/en/stable/user_guide/


## Nadchodzą problemy...

* W środowisku Pythona możemy mieć tylko jedną wersję danego pakietu, co może rodzić konflikty zależności:
    * np. pracujemy nad swoim projektem wykorzystującym najnowszą wersję pakietu A
    * i jednocześnie chcemy uruchomić projekt, w którym wymagana jest starsza wersja pakietu A.
* Na Linuksie i OSX Python jest używany przez system operacyjny. Pakiety (lub ich wersje), które doinstalujemy w głównym katalogu mogą zdestabilizować narzędzia systemowe...
* ...i odwrotnie, przy aktualizacji system może nam nadpisać wersje pakietów
* Możemy mieć problem z uporządkowaniem wymagań naszego projektu -- a chcielibyśmy móc uruchomić go nie tylko na naszym komputerze.

### Pytania do publiczności:
    
* (linux) czy zawsze potrzebujemy praw administratora (roota), żeby doinstalować pakiety do systemowego Pythona?
* (all) czy posprzątam swoje środowisko, jeśli odpalę `pip uninstall` dla każdego pakietu, który zainstalowałam przez `pip install`?
* (linux) czy mogę coś zepsuć wywołując `sudo pip install [paczka]`?
* (all) jak zainstalować pakiet A w wersji NIE NOWSZEJ niż 2.1?

## Środowisko wirtualne - moduł `venv`

`python -m venv [nazwa_enva]`

* Kopia środowiska Pythona, tworzona na żądanie
   * binarka `python`
   * biblioteka standardowa
   * `pip` + skrypty do aktywacji i dezaktywacji środowiska
   * instalujemy tylko pakiety, których potrzebujemy w konkretnym projekcie

* Izolacja od pozostałych środowisk 👍

### Sposób użycia - moduł `venv`

```console
$ python -m venv some/path/to/some_project_venv
```

Utworzy środowisko w zadanym katalogu.

```console
$ ls some_project_venv
bin  include  lib  lib64

$ ls -l some_project_venv/bin
-rw-r--r-- 1 asia asia 2248 11-25 22:05 activate
-rwxr-xr-x 1 asia asia  257 11-25 22:05 pip
lrwxrwxrwx 1 asia asia   11 11-25 22:05 python -> /bin/python
[...]

$ ls some_project_venv/lib/Python3.7
site-packages
```

### Aktywacja środowiska - skrypt shellowy `activate`


Platform  	|Shell 	|Command to activate virtual environment
---------------|----|-----------------------------------------
Posix 	|bash/zsh 	|`$ source <venv>/bin/activate`
  |  	fish 	|`$ . <venv>/bin/activate.fish`
  |  	csh/tcsh 	|`$ source <venv>/bin/activate.csh`
Windows 	|cmd.exe 	|`C:\> <venv>\Scripts\activate.bat`
  |  	PowerShell 	|`PS C:\> <venv>\Scripts\Activate.ps1`

https://docs.python.org/3/library/venv.html#creating-virtual-environments

### Jak to działa?

Zmiany w bieżącej sesji terminala:

* modyfikacja *zmiennej środowiskowej* `PATH` - Skierowanie na `pythona`, `pip`a i inne pliki wykonywalne pod nowym adresem
* ustawienie zmiennej środowiskowej `VIRTUAL_ENV` - po uruchomieniu Python wie, skąd ładować importowane pakiety, a pip wie, gdzie instalować
* stworzenie komendy (funkcji w shellu) `deactivate`


### Zmienne środowiskowe

Różne nazwane wartości, z których korzystają procesy uruchamiane w systemie operacyjnym. Gdy używamy terminala (shella=powłoki), możemy zmodyfikować zmienne środowiskowe tak, by wpłynąć na działanie programu, który za chwilę w tym shellu uruchomimy. Ze zmiennych środowiskowych korzysta zarówno Windows, Linuksy, jak i OSX.

Przykłady:

* `HOME` - adres katalogu domowego bieżącego użytkownika
* `USER` - nazwa bieżącego użytkownika
* `PATH` - lista katalogów, w których shell będzie szukał pliku wykonywalnego, gdy wpiszemy jakąś komendę bez podania pełnej ścieżki. Np. `python` :-)


### Przykład

```bash
# przed aktywacją virtual enva
$ echo $PATH
/home/asia/.local/bin:/home/asia/bin:/usr/local/bin:/usr/bin:/bin
$ which python
/bin/python
$ source some_project_venv/bin/activate
```

### Przykład c.d.

```bash
# po aktywacji
(some_project_venv) $ echo $PATH
/home/asia/some_project_venv/bin:/home/asia/.local/bin:/home/asia/bin:/usr/local/bin:/usr/bin:/bin
(some_project_venv) $ which python
/home/asia/some_project_venv/bin/python
(some_project_venv) $ echo $VIRTUAL_ENV
/home/asia/some_project_venv

(some_project_venv) $ deactivate
$ which python
/bin/python
```


## Jak tego używać? Przykład 1

Środowisko wirtualne w katalogu projektu:

* w katalogu projektu (powiedzmy `my_project`) odpalam `python -m venv env_my_project`
* [jeśli używam `git`a, wpisuję katalog `env_my_project` do pliku `.gitignore`]
* aktywuję środowisko `source env_my_project/bin/activate` (za każdym razem, gdy w nowym terminalu uruchamiam pliki projektu)
* instaluję potrzebne pakiety (to tylko raz)
* gdy projekt działa, zapisuję jego wymagania w pliku requirements: `pip freeze > requirements.txt`. Przyda się, gdy będę budować środowisko od nowa (`pip install -r requirements.txt`)
* gdy kończę pracę nad projektem i nie chcę zamykać okna terminala, dezaktywuję środowisko poleceniem `deactivate`
   

## Dobre rady

* NIE instalujemy pakietów w głównym środowisku Pythona. To się może źle skończyć.
* narzędzia, do których chcemy mieć dostęp w całym systemie (jako nasz użytkownik) instalujemy przez `pip install --user [paczka]`
* do każdego projektu w pythonie używamy osobnego virtualenva

### Pytania do publiczności

* czy mogę stworzyć środowisko wirtualne z wersją Pythona, której nie mam zainstalowanej w systemie?
* czy jeśli zainstaluję jakiś pakiet w wirtualnym środowisku, to będzie on dostępny dla głównego, systemowego Pythona?
* gdzie będzie 'widoczny' mój virtualenv, gdy go aktywuję?
* co się stanie, gdy w terminalu skasuję zawartość zmiennej środowiskowej `PATH`, np. poleceniem `export PATH=""`?

## Jak tego używać?  :-(

* :-( muszę ręcznie stworzyć i nazwać środowisko wirtualne
* :-( muszę pamiętać, jak mój virtualenv się nazywa i gdzie w systemie plików się znajduje
* :-( aby aktywować środowisko muszę wpisywać długie polecenia
* :-( moje środowisko wirtualne ma masę niepotrzebnych pakietów i nie jestem pewna/pewien, które dokładnie pakiety mogę usunąć

## Trochę prościej - `virtualenvwrapper`

https://virtualenvwrapper.readthedocs.io/en/latest/

* środowiska wirtualne w jednym miejscu
* wygodne polecenia, np. `mkvirtualenv [nazwa]`, `workon [nazwa]`
* możliwość skonfigurowania auto-aktywacji środowiska po wejściu do katalogu projektu


## `pipenv` - poziom wyżej

https://pipenv.readthedocs.io/en/latest/

Rekomendowany system zarządzania zależnościami w projektach pythonowych.

Używa `pip` i `venv` za nas, dodając do tego wiele mądrych automatyzacji.

### `pipenv` - szybki start

https://pipenv.readthedocs.io/en/latest/install/#installing-pipenv
    
OSX: `brew install pipenv`

Win/Linux: `pip install --user pipenv`

### `pipenv` - użycie w nowym projekcie

* przechodzimy w konsoli do katalogu projektu: `cd my_project`
* od razu coś instalujemy, np. bibliotekę `requests`: `pipenv install requests`
* `pipenv` automatycznie znajduje/tworzy środowisko wirtualne (w swoim centralnym katalogu)
* automatycznie tworzy w projekcie pliki `Pipfile` i `Pipfile.lock` (lepszy odpowiednik `requirements.txt`)



```console
➜  experiments$ mkdir my_project
➜  experiments$ cd my_project 
➜  my_project$ pipenv install requests
Creating a virtualenv for this project…
Pipfile: /home/asia/experiments/my_project/Pipfile
Using /usr/bin/python (3.7.1) to create virtualenv…
✔ Complete 
Using interpreter /bin/python
New python executable in /home/asia/venvs/my_project-F8Fo0uYg/bin/python
Installing setuptools, pip, wheel...done.
Virtualenv location: /home/asia/venvs/my_project-F8Fo0uYg
Creating a Pipfile for this project…
Installing requests…
Adding requests to Pipfile's [packages]…
✔ Installation Succeeded 
Pipfile.lock not found, creating…
Locking [dev-packages] dependencies…
Locking [packages] dependencies…
✔ Success! 
Updated Pipfile.lock (444a6d)!
Installing dependencies from Pipfile.lock (444a6d)…
  🐍   ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 5/5 — 00:00:01
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.

```

### `pipenv` - aktywacja środowiska

* aktywujemy środowisko poleceniem `pipenv shell`, deaktywujemy poleceniem `exit` (uruchamiamy/kończymy nowy shell)
* możemy też uruchomić swoje programy bez wcześniejszej aktywacji, np. `pipenv run python my_program.py`

### `pipenv` - Pipfile + Pipfile.lock

* `Pipfile` - deklarowane zależności projektu (np. `requests`) - czytelny dla człowieka
* każda instalacja/deinstalacja aktualizuje `Pipfile`
* `Pipfile.lock` - wszystkie zależności (czyli np. także zależności biblioteki `requests`), stan całego środowiska - czytelny dla pipenva

### `pipenv` - install

`$ pipenv install` - spowoduje zainstalowanie wszystkiego co jest opisane w pliku `Pipfile.lock`

`$ pipenv sync` - spowosuje zainstalowanie wszystkiego, bez ponownego lokowanie `Pipfile.lock`

### Prosty `Pipfile`

```toml
[[source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[dev-packages]

[packages]
requests = "*"

[requires]
python_version = "3.7"
```

### `pipenv` - Sprzątanie środowiska

`$ pipenv uninstall [paczka]` - usuwa pakiet, aktualizuje Pipfile


`$ pipenv clean` - usunie niepotrzebne zależności pakietów, których już nie ma w Pipfile


### Ja chcę tylko zainstalować narzedzie...

* Na ratunek przychodzi pipX https://github.com/pipxproject/pipx  
* Jest to narzędzie które instaluje każdy pakiet w dedykowanym wirutalnym środowisku
* Dzięki temu możemy w prosty sposób zainstalować narzędzia napisane w pythonie. Bez martwienia się o problemy z wymaganiami i integralnością naszego bazowego środowiska

### Pytania do publiczności

* jak się ma `pip` + `venv` do `pipenv`?
* jakich narzędzi będę używać do zarządzania zależnościami w swoim projekcie i czemu to będzie `pipenv`?

## Ile Pythonów zmieści się w jednym systemie?


Wiele. Możemy mieć zainstalowanych kilka wersji Pythona (np. 2.7.x, 3.6.x oraz 3.7.x), a nawet kilka dystrybucji tej samej wersji (Anaconda oraz oficjalny Python https://python.org ).



Zarządzanie wersjami Pythona - do poczytania:
* https://github.com/pyenv/pyenv
* https://github.com/asdf-vm/asdf

## Podsumowanie

* po co nam środowiska wirtualne?
   * izolacja projektów 
       * różne wersje zależności
       * niemodyfikowanie Pythona systemowego
   * powtarzalność instalacji
   * przewidywalność działania oprogamowania - pakiety mogą wpływać na działanie innych pakietów
   
* jakie mamy narzędzia?
   * wbudowane w Pythona - `pip` i `venv`
   * ładne opakowania do powyższych - `virtualenvwrapper`, a przede wszystkim `pipenv`
   
* dobre praktyki:
   * używać środowisk wirtualnych, **zawsze**!

Koniec części 1 :)